FlowCal/mef.py

```diff
@@ -265,26 +319,30 @@ def fit_beads_autofluorescence(fl_channel, fl_mef):
     # Initial guesses:
     # 0: slope found by putting a line through the highest two points.
     # 1: y-intercept found by putting a line through highest two points.
-    # 2: bead autofluorescence initialized to 100.
+    # 2: bead autofluorescence initialized using the first point.
     params[0] = (np.log(fl_mef[-1]) - np.log(fl_mef[-2])) / \
-                    (fl_channel[-1] - fl_channel[-2])
-    params[1] = np.log(fl_mef[-1]) - params[0] * fl_channel[-1]
-    params[2] = 100.
+                (np.log(fl_au[-1]) - np.log(fl_au[-2]))
+    params[1] = np.log(fl_mef[-1]) - params[0] * np.log(fl_au[-1])
+    params[2] = np.exp(params[0]*np.log(fl_au[0]) + params[1]) - fl_mef[0]
 
     # Error function
     def err_fun(p, x, y):
-        return np.sum((np.log(y + p[2]) - ( p[0] * x + p[1] ))**2)
-        
+        return np.sum((np.log(y + p[2]) - ( p[0] * np.log(x) + p[1] ))**2)
+
     # Bead model function
     def fit_fun(p,x):
-        return np.exp(p[0] * x + p[1]) - p[2]
+        return np.exp(p[0] * np.log(x) + p[1]) - p[2]
 
-    # Channel-to-MEF standard curve transformation function
+    # a.u.-to-MEF standard curve transformation function
+    # We use the equivalent form ``exp(p[1]) * x**p[0]``. Mathematically, this
+    # is undefined for x < 0. In this case, we redefine the output as
+    # ``- exp(p[1]) * (-x)**p[0]``. These two can be combined as follows:
+    # ``np.sign(x) * np.exp(p[1]) * (np.abs(x)**p[0])``.
     def sc_fun(p,x):
-        return np.exp(p[0] * x + p[1])
+        return np.sign(x) * np.exp(p[1]) * (np.abs(x)**p[0])
```

Comment:
>  i imagine this is just a rearrangement of the bead model, minus the autofluorescence term?

Yes.

> you can raise a negative number to an exponent...

Not really... http://math.stackexchange.com/questions/415141/is-there-a-problem-when-defining-exponential-with-negative-base. In fact, the reason why I used this "extended" equation is because the numpy functions involved were initially failing for negative values.

I guess a more formal way to do it is the following. If `x<0`:

```
x**p0 = exp(Log(x))**p0
x**p0 = exp(Log(abs(x)) + i*pi)**p0
x**p0 = (exp(log(abs(x))) * exp(i*pi))**p0
x**p0 = (exp(log(abs(x))) * exp(i*pi))**p0
x**p0 = (abs(x)**p0) * (-1)**p0
```

Where I'm using the definition of complex logarithm (https://en.wikipedia.org/wiki/Complex_logarithm). In the last equation, `(-1)**po` is still not well defined, so I can't think of a 100% formally correct solution. Equivalently,

```
x**p0 = (abs(x)**p0) * exp(i*p0*pi)
```

I'm relying on `p0` being close to one (standard curve is close to a line) to approximate this as `exp(i*pi) = -1`. If it's not one, there might be some distortion applied to negative values because `exp(i*p0*pi)` is complex. We can still justify this equation by saying that we're using the negative absolute value of `exp(i*p0*pi)` (which is `-1`). Using the negative guarantees that negative numbers in a.u. remain negative. If we choose to use the real part of `exp(i*p0*pi)`, i.e. `cos(p0*pi)`, this value will be between `-1` and `1`, and I don't think this is appropriate.

Again, this method relies on `p0` being close to 1. This is not perfect, but all the floating-point data in my test dataset has given a `p0` that is very close to one, up to 1e-2 error. If you have a better idea on how to handle this, let me know. What we **can't** do is ignore or discard negative values. This is why we have this code review stage, as I didn't anticipate this issue until I started writing code.

If this makes sense, I'll update the comments to reflect this (better) description.